src/NServiceBus.Transport.Msmq/MessagePump.cs

```diff
@@ -9,69 +9,86 @@ namespace NServiceBus.Transport.Msmq
+            var inputAddress = MsmqAddress.Parse(receiveSettings.ReceiveAddress);
+            var errorAddress = MsmqAddress.Parse(receiveSettings.ErrorQueue);
 
-            if (!string.Equals(inputAddress.Machine, RuntimeEnvironment.MachineName, StringComparison.OrdinalIgnoreCase))
+            if (!string.Equals(inputAddress.Machine, RuntimeEnvironment.MachineName,
+                StringComparison.OrdinalIgnoreCase))
             {
-                throw new Exception($"MSMQ Dequeuing can only run against the local machine. Invalid inputQueue name '{settings.InputQueue}'.");
+                throw new Exception(
+                    $"MSMQ Dequeuing can only run against the local machine. Invalid inputQueue name '{receiveSettings.ReceiveAddress}'.");
             }
 
             inputQueue = new MessageQueue(inputAddress.FullPath, false, true, QueueAccessMode.Receive);
             errorQueue = new MessageQueue(errorAddress.FullPath, false, true, QueueAccessMode.Send);
 
-            if (settings.RequiredTransactionMode != TransportTransactionMode.None && !QueueIsTransactional())
+            if (transportSettings.TransportTransactionMode != TransportTransactionMode.None && !QueueIsTransactional())
             {
-                throw new ArgumentException($"Queue must be transactional if you configure the endpoint to be transactional ({settings.InputQueue}).");
+                throw new ArgumentException(
+                    $"Queue must be transactional if you configure the endpoint to be transactional ({receiveSettings.ReceiveAddress}).");
             }
 
             inputQueue.MessageReadPropertyFilter = DefaultReadPropertyFilter;
 
-            if (settings.PurgeOnStartup)
-            {
-                inputQueue.Purge();
-            }
-
-            receiveStrategy = receiveStrategyFactory(settings.RequiredTransactionMode);
-
-            receiveStrategy.Init(inputQueue, errorQueue, onMessage, onError, criticalError, discardExpiredTtbrMessages);
+            receiveStrategy = receiveStrategyFactory(transportSettings.TransportTransactionMode);
+            receiveStrategy.Init(inputQueue, errorQueue, onMessage, onError, criticalErrorAction, transportSettings.IgnoreIncomingTimeToBeReceivedHeaders);
 
+            maxConcurrency = limitations.MaxConcurrency;
+            concurrencyLimiter = new SemaphoreSlim(limitations.MaxConcurrency, limitations.MaxConcurrency);
             return TaskEx.CompletedTask;
         }
 
-        public void Start(PushRuntimeSettings limitations)
+        public Task StartReceive()
         {
             MessageQueue.ClearConnectionCache();
 
-            maxConcurrency = limitations.MaxConcurrency;
-            concurrencyLimiter = new SemaphoreSlim(limitations.MaxConcurrency, limitations.MaxConcurrency);
-            cancellationTokenSource = new CancellationTokenSource();
+            if (receiveSettings.PurgeOnStartup)
+            {
+                inputQueue.Purge();
```

Comment:
Why have we moved the purge? Does it change the bahavior in any way?

Reply:
I was thinking it should probably only purge once you start to receive the queue, not when you define the queues already. But thinking a bit more about it, maybe we need to purge earlier. In core, this is actually important as feature startup tasks (before transport initialize & start) could send subscription messages, which would now be purged once you start the receiver.

So in essence, purging would have to go pack to when we initialize the transport, which I dislike a bit but I guess it needs to behave that way for compatibility reasons.

Reply:
I moved the purging to Initialize, but due to the current usage of the API by core, this doesn't really change the behavior at all. This requires https://github.com/Particular/NServiceBus/pull/5906 in addition to make the behavior similar to v7